### PR TITLE
XRT-932/930 enhance logging and forward to uart

### PR DESCRIPTION
**Pull request by xdavidz**
#### Problem solved by the commit
1) Enabling RPU logging mechanism
2) Enabling UART lite for RPU
3) Reporting RPU status
4) Renaming xgq_xocl in xgq driver

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Introduce new op code (vmr control) to retrieve vmr status.
RPU log will be redirected to UART lite and shared memory.
When system has timeout, logs cached in shared memory will be dumped to host side debug fs "/sys/kernel/debug/xclmgmt/trace"
User can change RPU log level by "echo 0-2" into sysfs.
Change xgq_xocl to another name that is not being conflicted with shared xgq common code that is introduced by @mamin506  

#### Documentation impact (if any)
https://confluence.xilinx.com/pages/viewpage.action?pageId=302345261#XGQ1.0CommandandQueueSpecification(WIP)-Overview

### CODE REVIEW

**Review by maxzhen — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_fdt.h [hunk 1]
```diff
@@ -113,7 +113,7 @@
 #define NODE_PCIE "pcie"
 #define NODE_BARS "bars"
 #define NODE_XGQ_SQ_BASE "ep_xgq_mgmt_to_rpu_sq_pi_00"
-#define NODE_XGQ_RING_BASE "ep_xgq_payload_mgmt_00"
+#define NODE_XGQ_PAYLOAD_BASE "ep_xgq_payload_mgmt_00"
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/xocl_fdt.h:116]: Is NODE_XGQ**_VMR**_PAYLOAD_BASE better?

File: src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.cpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2021 Xilinx, Inc
+ * Copyright (C) 2022 Xilinx, Inc
```
Comment (mamin506) [inline src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.cpp:2]: 2021-2022

File: src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.h [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2021 Xilinx, Inc
+ * Copyright (C) 2022 Xilinx, Inc
```
Comment (mamin506) [inline src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.h:2]: 2021-2022

**Review by mamin506 — CHANGES_REQUESTED** [review]
The renaming change looks good to me.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 4]
```diff
@@ -293,6 +294,31 @@ static bool xgq_submitted_cmds_empty(struct xocl_xgq *xgq)
 	return false;
 }
 
+static void xgq_vmr_log_dump(struct xocl_xgq_vmr *xgq)
+{
+	struct vmr_log log;
+
+	memcpy_fromio(&xgq->xgq_vmr_shared_mem, xgq->xgq_payload_base,
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:301]: could you use xocl_memcpy_fromio()?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 5]
```diff
@@ -293,6 +294,31 @@ static bool xgq_submitted_cmds_empty(struct xocl_xgq *xgq)
...
+		sizeof(xgq->xgq_vmr_shared_mem));
+
+	if (xgq->xgq_vmr_shared_mem.vmr_magic_no == VMR_MAGIC_NO) {
+		u32 idx, log_idx = xgq->xgq_vmr_shared_mem.log_msg_index;
+
+		XGQ_DBG(xgq, "=== start dumping vmr log ===");
+		for (idx = 0; idx < VMR_LOG_MAX_RECS; idx++) {
+			memcpy_fromio(&log.log_buf, xgq->xgq_payload_base +
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:309]: xocl_memcpy_fromio()

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 6]
```diff
@@ -445,66 +473,66 @@ static int submit_cmd(struct xocl_xgq *xgq, struct xocl_xgq_cmd *cmd)
...
  */
-static u64 memcpy_to_devices(struct xocl_xgq *xgq, const void *xclbin_data,
+static u64 memcpy_to_devices(struct xocl_xgq_vmr *xgq, const void *xclbin_data,
 	size_t xclbin_len)
 {
-	void __iomem *dst = xgq->xgq_ring_base + XOCL_XGQ_DATA_OFFSET;
+	u32 offset = xgq->xgq_vmr_shared_mem.vmr_data_start;
+	void __iomem *dst = xgq->xgq_payload_base + offset;
 
 	memcpy_toio(dst, xclbin_data, xclbin_len);
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:496]: xocl_memcpy_toio()

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_fdt.h (hunk 1 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/xocl_fdt.h:116]: yes, fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 4 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:301]: yes, fixed

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 5 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:309]: yes, fixed

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_fdt.h (hunk 1 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/xocl_fdt.h]: yes, fixed.

File: src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.cpp (hunk 2 above)
Comment (xdavidz) [reply src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.cpp]: yes, fixed.

File: src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.h (hunk 3 above)
Comment (xdavidz) [reply src/runtime_src/core/tools/xbmgmt2/flash/xgq_vmr.h]: thanks, fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 4 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c]: fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 5 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c]: fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 6 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:502]: fixed.

**Review by houlz0507 — APPROVED** [review]

**Review by mamin506 — APPROVED** [review]

**Review by stsoe — APPROVED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c [hunk 7]
```diff
@@ -293,6 +294,31 @@ static bool xgq_submitted_cmds_empty(struct xocl_xgq *xgq)
 	return false;
 }
 
+static void xgq_vmr_log_dump(struct xocl_xgq_vmr *xgq)
+{
+	struct vmr_log log;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c:299]: Please initialize this structure to zeros.
```suggestion
	struct vmr_log log = {0};
```
Note: the '0' character is not needed, but having it helps defined what the intention is.

**Review by rozumx — APPROVED** [review]
Approved with 1 minor structure initialization issues.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c (hunk 7 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c]: fixed, thanks.

**Pull request merged**